src/modules/schnorrsig/main_impl.h

```
@@ -264,4 +264,206 @@ int secp256k1_schnorrsig_verify(const secp256k1_context* ctx, const unsigned cha
+    /* where n' = n_before                                              */
+    secp256k1_schnorrsig_sha256_tagged_aggregation(&hash);
+    for (i = 0; i < n_before; ++i) {
+        /* serialize pk_i */
+        unsigned char pk_ser[32];
+        if (!secp256k1_xonly_pubkey_serialize(ctx, pk_ser, &all_pubkeys[i])) {
+            return 0;
+        }
+        /* write r_i */
+        secp256k1_sha256_write(&hash, &aggsig[i*32], 32);
+        /* write pk_i */
+        secp256k1_sha256_write(&hash, pk_ser, 32);
+        /* write mes_i*/
+        secp256k1_sha256_write(&hash, &all_msgs32[i*32], 32);
+    }
+
+    /* Compute s = s_old + sum_{i = n_before}^{n} z_i*s_i */
+    /* where s_old = 0 if n_before = 0 */
+    secp256k1_scalar_set_int(&s, 0);
+    if (n_before > 0) {
+        secp256k1_scalar_set_b32(&s, &aggsig[n_before*32], &overflow);
+        if (overflow) {
+            return 0;
+        }
+    }
+    for (i = n_before; i < n; ++i) {
+        unsigned char pk_ser[32];
+        unsigned char hashoutput[32];
+        secp256k1_sha256 hashcopy;
+        secp256k1_scalar si;
+        secp256k1_scalar zi;
+
+        /* Step 0: Serialize pk_i into pk_ser   */
+        if (!secp256k1_xonly_pubkey_serialize(ctx, pk_ser, &all_pubkeys[i])) {
+            return 0;
+        }
+
+        /* Step 1: z_i = TaggedHash(...) */
+        /* 1.a) Write into hash r_i, pk_i, mes_i, r_i */
+        secp256k1_sha256_write(&hash, &new_sigs64[(i-n_before)*64], 32);
+        secp256k1_sha256_write(&hash, pk_ser, 32);
+        secp256k1_sha256_write(&hash, &all_msgs32[i*32], 32);
+        /* 1.b) Copy the hash */
+        hashcopy = hash;
+        /* 1.c) Finalize the copy to get zi*/
+        secp256k1_sha256_finalize(&hashcopy, hashoutput);
+        secp256k1_scalar_set_b32(&zi, hashoutput, NULL);
+
+        /* Step2: s := s + zi*si */
+        secp256k1_scalar_set_b32(&si, &new_sigs64[(i-n_before)*64+32], NULL);
```

I wonder if this should really ignore the overflow. Overflow should never happen here for a valid input sig, so it's probably stricter to bail out. (I see that this will be a change to the spec. cc @jonasnick )

---

i think when i wrote this, i liked the idea of having no error paths. currently it returns 1 always (if the overflow of the previous s above is fixed to be consistent with the spec). however, failing early also has it's merit. i opened a pr for the spec change: https://github.com/blockstreamresearch/cross-input-aggregation/pull/10.